shiny/_template_utils.py

```diff
@@ -67,20 +71,99 @@ def template_query(question_state: Optional[str] = None):
     elif template in package_template_choices.values():
         js_component_questions(template)
     else:
-        app_template_questions(template)
+        app_template_questions(template, mode)
 
 
-def app_template_questions(template: str):
+def download_and_extract_zip(url: str, temp_dir: Path):
+    response = requests.get(url)
+    response.raise_for_status()
+    zip_file_path = temp_dir / "repo.zip"
+    zip_file_path.write_bytes(response.content)
+    with zipfile.ZipFile(zip_file_path, "r") as zip_file:
+        zip_file.extractall(temp_dir)
+
+
+def use_git_template(url: str, mode: Optional[str] = None):
+    # Github requires that we download the whole repository, so we need to
+    # download and unzip the repo, then navigate to the subdirectory.
+
+    parsed_url = urlparse(url)
+    path_parts = parsed_url.path.strip("/").split("/")
+    repo_owner, repo_name, _, branch_name = path_parts[:4]
+    subdirectory = "/".join(path_parts[4:])
+
+    zip_url = f"https://github.com/{repo_owner}/{repo_name}/archive/refs/heads/{branch_name}.zip"
+
+    with tempfile.TemporaryDirectory() as temp_dir:
```

Comment:
Does this auto delete `temp_dir` when the `with` context exits?

Reply:
It does, I know this because I was returning outside of the with context for a while and nothing worked.